docs/proposals/28-autonomous-shoot-clusters.md

```diff
@@ -0,0 +1,495 @@
+This simplifies the crafting of these documents, and it also ensures that the configuration and versions used for the bootstrapping of the autonomous shoot cluster fit to the configuration later found in the connected Gardener system.
+
+#### `gardenadm init`
+
+This command bootstraps the first control plane node of the autonomous shoot cluster.
+The configuration is specified via separate files.
+It needs a `Shoot` manifest and some additional configuration files, which usually reside in the garden cluster, e.g., a `CloudProfile` (best downloaded with `gardenadm discover`).
+For the initial cluster Gardener bootstrapping use-case however, these won't come from a pre-existing garden cluster as it doesn't exist yet.
+They can then be provided locally (that's why `gardenadm discover` and `gardenadm init` are separate commands), reflecting the to-be-deployed Gardener system later.
+
+The high-level steps executed by the command are as follows:
+
+1. Generate certificate authorities, certificates and keys using the [secrets manager](../development/secrets_management.md).
+2. Create [`OperatingSystemConfig` resource](../extensions/operatingsystemconfig.md) containing `systemd` units and files.
+   This includes the files for the minimal standard Kubernetes control plane, i.e., `kubelet`, `etcd`, `kube-apiserver`, `kube-controller-manager`, and `kube-scheduler`.
+4. Apply the `OperatingSystemConfig` resource to the node using the [`gardener-node-agent`](../concepts/node-agent.md).
+5. `kubelet` starts and runs the static pods for the minimal standard Kubernetes control plane (`etcd`, `kube-apiserver`, `kube-controller-manager` and `kube-scheduler`).
+6. Deploy [`gardener-resource-manager`](../concepts/resource-manager.md) in host network mode.
+7. Apply the corresponding provider and networking extensions in host network mode.
+8. Deploy `kube-proxy` and CoreDNS for service routing and domain name resolution.
+9. Create a [`Network` resource](../extensions/network.md) and apply it to generate `ManagedResource` associated with the pod network.
+10. Apply the corresponding provider and networking extensions in pod network mode.
+11. Deploy `gardener-resource-manager` in pod network mode.
+12. Regenerate `OperatingSystemConfig` and activate `gardener-node-agent`.
+13. Apply [`ControlPlane` resource](../extensions/controlplane.md).
+
+##### Problems & Solutions
+
+As we intend to re-use code as much as possible, we need to jump through some hoops to make this work.
+
+###### Faking a Kubernetes API
+
+Most of the Gardener code base assumes an always existing garden cluster or the API server of a seed cluster, which might not be available during bootstrapping.
+A somewhat simple solution for components running as part of the same process is to use a fake client, e.g., `sigs.k8s.io/controller-runtime/pkg/client/fake` or `github.com/gardener/gardener/pkg/client/kubernetes/fake`.
```

Comment:
> components running as part of the same process

What about components that are not running in the same process?

Reply:
If components require to be run as separate processes or the integration into a single process is too much effort, the second best option is to run a simple/small kubernetes cluster, e.g. KinD. We propose to use this approach in the medium touch scenario (see
https://github.com/gardener/gardener/blob/68ccf134a903568bb724f0926d5c3dfa2e470379/docs/proposals/28-autonomous-shoot-clusters.md#gardenadm-bootstrap) as we deem running MCM and a provider extension in the same process might be too much effort. For the initial control plane node bootstrap it seems to be possible to have everything in one process, though.
Have you some requirements in mind that would require a second process with access to an API server?

Reply:
`gardenadm` will run as a single process and include the initial steps to roll out the initial kubernetes control plane. This includes the initial roll out of the operating system configuration by reusing the node agent (go) functions. For this, the fake API seems sufficient.
As soon as other components, e.g. `gardener-resource-manager` or extensions, come into play they connect to the initial kubernetes control plane exposed via the running static pods.
To summarise: The fake API is only used initially during the bootstrap process. As soon as the kubernetes control plane is up and running, it will be used instead.